main.go

```diff
@@ -0,0 +1,543 @@
+	//       <TimeNextVisible>...</TimeNextVisible>
+	//       <DequeueCount>...</DequeueCount>
+	//       <MessageText>...</MessageText>
+	//     </QueueMessage>
+	//     ...
+	// </QueueMessagesList>
+	// ```
+	// We unmarshal the response into go objects, using the go xml decoder.
+	fullBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	reader := strings.NewReader(string(fullBody))
+	dec := xml.NewDecoder(reader)
+	err = dec.Decode(&queueMessagesList)
+	if err != nil {
+		return nil, err
+	}
+	if len(queueMessagesList.QueueMessages) == 0 {
+		return nil, errors.New("Zero tasks returned in Azure XML QueueMessagesList")
+	}
+	if size := len(queueMessagesList.QueueMessages); size > 1 {
+		return nil, fmt.Errorf("%v tasks returned in Azure XML QueueMessagesList, even though &numofmessages=1 was specified in poll url", size)
+	}
+
+	// at this point we know there is precisely one QueueMessage (== task)
+	qm := queueMessagesList.QueueMessages[0]
+
+	// To find the task referenced in a message the worker must base64
+	// decode and JSON parse the contents of the <MessageText> tag. This
+	// would return an object on the form: {taskId, runId}.
+	m, err := base64.StdEncoding.DecodeString(qm.MessageText)
+	if err != nil {
+		return nil, err
+	}
+
+	// Utility method for replacing a placeholder within a uri with
+	// a string value which first must be uri encoded...
+	detokeniseUri := func(uri, placeholder, rawValue string) string {
+		return strings.Replace(uri, placeholder, strings.Replace(url.QueryEscape(rawValue), "+", "%20", -1), -1)
+	}
+
+	// Before using the signedDeleteUrl the worker must replace the placeholder
+	// {{messageId}} with the contents of the <MessageId> tag. It is also
+	// necessary to replace the placeholder {{popReceipt}} with the URI encoded
+	// contents of the <PopReceipt> tag.  Notice, that the worker must URI
+	// encode the contents of <PopReceipt> before substituting into the
+	// signedDeleteUrl. Otherwise, the worker will experience intermittent
+	// failures.
+
+	// Since urlPair is a value, not a pointer, we can update this copy which
+	// is associated only with this particular task
+	urlPair.SignedDeleteUrl = detokeniseUri(
+		detokeniseUri(
+			urlPair.SignedDeleteUrl,
+			"{{messageId}}",
+			qm.MessageId,
+		),
+		"{{popReceipt}}",
+		qm.PopReceipt,
+	)
+
+	// initialise fields of TaskRun not contained in json string m
+	taskRun := TaskRun{
+		QueueMessage:  qm,
+		SignedURLPair: urlPair,
+	}
+	// now populate remaining json fields of TaskRun from json string m
+	err = json.Unmarshal(m, &taskRun)
+	if err != nil {
+		return nil, err
+	}
+
+	// Workers should read the value of the `<DequeueCount>` and log messages
+	// that alert the operator if a message has been dequeued a significant
+	// number of times, for example more than 10.
+	if qm.DequeueCount > 10 {
+		log.Printf("WARN: Task with id %v has been dequeued %v times!", taskRun.TaskId, qm.DequeueCount)
```

Comment:
You must call `queue.claimTask` before deleting the message... So definitely retry the delete operation.

Basically not retrying a request with exponential backoff an extreme corner case that should only be used in very very rare cases... Even if exponential backoff slows down the request rate that can be important so the server has time to recover.

For something like a service that runs on heroku that has to return API calls within 30s, we like to make retries quick... still exponential, but very quick... I think it's up to 6s sleep or so... While things that run as background workers should generally sleep longer for retries... and definitely retry. It improves robustness and reduces obscurity.
So even in cases where you can argue that it's not critical, you might as well just to keep a good pattern going :)


> I wonder if maybe I should put the retry support directly in taskcluster-client-go library...
If `taskcluster-client-go` doesn't implement retries with exponential backoff that is a bug.
Client retry logic is implemented here: https://github.com/taskcluster/taskcluster-client/blob/master/lib/client.js#L292-L303

With the following configuration:
https://github.com/taskcluster/taskcluster-client/blob/master/lib/client.js#L46-L55

Using `delayFactor:    100,` is the same as AWS, and great for servers that needs to reply within 30s, but for background processes and workers I would recommend 500ms instead... There is no reason a worker shouldn't be patient.

If workers are too eager they could overload the servers and if backoff isn't exponential they would able to keep the servers down. So `delayFactor:    500,` is probably a good idea for a background worker, as it gives delays: 500ms, 1s, 2s, 4s, 8s, 16s (with 5 retries). Such delays should give the servers some air so that they have a chance of getting on top. Note, we typically do 1 requests followed by up to 5 retries by default, and default to `delayFactor:    100,`, but encourage background processes and workers to use `delayFactor:    500,`.

Funny story when pulse was down for 10 min, queue.taskcluster.net experienced 50k failed requests... Probably because of retries. Because normally we don't see that many requests, anyways when that many clients keep retrying it's nice that retries are exponential, or servers might crash as soon as they come online again.

Note, one might even argue that we should implement some random factor of delay, i.e multiply the delay for each retry as follows `delay = delay * (Math.random() * 0.2 + 0.9)´.

----
@pmoore, I realize that this is not important until we have a lot of generic workers running. So I'm mostly emphasising the importance of retry logic, to ensure that we have good practices on this subject.